### PR TITLE
feat(api): Document dashboard details endpoint

### PR DESCRIPTION
**Pull request by narsaynorath**
Documents the dashboard details endpoint and makes GET, PUT, DELETE methods public API.

I had to rename the model serializer, which had the same name as the DRF serializer, to avoid build issues with the API docs.

### CODE REVIEW

File: src/sentry/api/serializers/models/dashboard.py [hunk 1]
```diff
@@ -166,8 +188,28 @@ def serialize(self, obj, attrs, user, **kwargs):
...
 
+class DashboardFilters(TypedDict):
+    release: list[str]
+
+
+class DashboardDetailsResponse(TypedDict):
+    id: str
+    title: str
+    dateCreated: str
+    createdBy: list[OpaqueSerializedResponse]
```
Comment (narsaynorath) [inline src/sentry/api/serializers/models/dashboard.py:199]: @sentaur-athena I wasn't sure how to type this `createdBy` field. It's derived from [this line of code](https://github.com/getsentry/sentry/blob/master/src/sentry/api/serializers/models/dashboard.py#L197) which makes a request to the user service and it lists `list[OpaqueSerializedResponse]` as the return of the `serialize_many` call.

This is supposed to be a user object. Is there an example of a place where this type of data is typed properly?

File: src/sentry/apidocs/build.py [hunk 2]
```diff
@@ -104,6 +104,16 @@ def get_old_json_components(filename: str) -> Any:
             "url": "https://github.com/getsentry/sentry-docs/issues/new/?title=API%20Documentation%20Error:%20/api/integration-platform/&template=api_error_template.md",
         },
     },
+    {
+        "name": "Dashboards",
+        "x-sidebar-name": "Dashboards",
+        "description": "Endpoints for Dashboards",
+        "x-display-description": True,
```
Comment (edwardgou-sentry) [inline src/sentry/apidocs/build.py:111]: Hmm, what does this attribute do? I noticed a majority of the tags have this  set to false instead.

**Review by edwardgou-sentry — APPROVED** [review]

File: src/sentry/apidocs/build.py (hunk 2 above)
Comment (narsaynorath) [reply src/sentry/apidocs/build.py:111]: Screenshot 2024-07-17 at 11 36 25 AM
it displays this description text. I'm going to remove it here.

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 3]
```diff
@@ -44,24 +54,28 @@ def _get_dashboard(self, request: Request, organization, dashboard_id):
...
+            404: RESPONSE_NOT_FOUND,
+        },
+        examples=DashboardExamples.DASHBOARD_GET_RESPONSE,
+    )
     def get(self, request: Request, organization, dashboard) -> Response:
         """
-        Retrieve an Organization's Dashboard
-        ````````````````````````````````````
-
         Return details on an individual organization's dashboard.
```
Comment (sentaur-athena) [inline src/sentry/api/endpoints/organization_dashboard_details.py:78]: noob question: Do customers know what Dashboard is? If it needs explanation can you also add that or link to product docs?

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 4]
```diff
@@ -105,17 +121,22 @@ def delete(self, request: Request, organization, dashboard) -> Response:
...
+        },
+        examples=DashboardExamples.DASHBOARD_PUT_RESPONSE,
+    )
     def put(self, request: Request, organization, dashboard) -> Response:
         """
-        Edit an Organization's Dashboard
-        ```````````````````````````````````
-
         Edit an individual organization's dashboard as well as
         bulk edits on widgets (i.e. rearranging widget order).
```
Comment (sentaur-athena) [inline src/sentry/api/endpoints/organization_dashboard_details.py:139]: What else can be done by editing widgets. Can you please expand on this?

File: src/sentry/apidocs/examples/dashboard_examples.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+from drf_spectacular.utils import OpenApiExample
+
+DASHBOARD_OBJECT = {
+    "id": "1",
+    "title": "Dashboard",
+    "dateCreated": "2024-06-20T14:38:03.498574Z",
+    "createdBy": {
```
Comment (sentaur-athena) [inline src/sentry/apidocs/examples/dashboard_examples.py:7]: What I understand is that anyone can fetch the dashboard, and that would give them all the details of `Admin`. Is that right? Should we revise this permission? Maybe we shouldn't include a user object but maybe just email.

**Comment by sentaur-athena** [comment]
@vivianyentran can you please review the docs? Thank you.

File: src/sentry/api/endpoints/organization_dashboard_details.py (hunk 4 above)
Comment (narsaynorath) [reply src/sentry/api/endpoints/organization_dashboard_details.py]: I expanded a bit more on the bulk edits here

File: src/sentry/api/endpoints/organization_dashboard_details.py (hunk 3 above)
Comment (narsaynorath) [reply src/sentry/api/endpoints/organization_dashboard_details.py]: I clarified by using the term "Custom Dashboard" in these docs. Customers should know if they're interested in this API, and it is a sidebar item as well

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 6]
```diff
@@ -71,17 +85,19 @@ def get(self, request: Request, organization, dashboard) -> Response:
...
+            404: RESPONSE_NOT_FOUND,
+        },
+    )
     def delete(self, request: Request, organization, dashboard) -> Response:
         """
-        Delete an Organization's Dashboard
-        ```````````````````````````````````
-
-        Delete an individual organization's dashboard, or tombstone
+        Delete an individual organization's custom dashboard, or tombstone
```
Comment (lizokm) [inline src/sentry/api/endpoints/organization_dashboard_details.py:99]:
```suggestion
        Delete an organization's custom dashboard, or tombstone
```

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 7]
```diff
@@ -105,17 +121,23 @@ def delete(self, request: Request, organization, dashboard) -> Response:
...
-
-        Edit an individual organization's dashboard as well as
-        bulk edits on widgets (i.e. rearranging widget order).
-
-        :pparam Organization organization: the organization the dashboard belongs to.
-        :pparam Dashboard dashboard: the old dashboard object
-        :auth: required
+        Edit an individual organization's custom dashboard as well as
+        bulk edits on widgets (i.e. rearranging widget order, updating
+        queries and fields, changing display types, etc).
```
Comment (lizokm) [inline src/sentry/api/endpoints/organization_dashboard_details.py:140]:
```suggestion
Edit an organization's custom dashboard as well as any bulk edits on widgets that may have been made. (For example, a custom widget order, updated queries and fields, specific display types, and so on.)
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 8]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
...
+        required=False, allow_null=True, help_text="The saved start time for this Dashboard."
+    )
+    end = serializers.DateTimeField(
+        required=False, allow_null=True, help_text="The saved end time for this Dashboard."
+    )
+    filters = serializers.DictField(
+        required=False, help_text="The saved filters for this Dashboard."
+    )
+    utc = serializers.BooleanField(
+        required=False, help_text="Setting for the saved time range of the Dashboard to be in UTC."
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:482]:
```suggestion
        required=False, help_text="Setting that lets you display saved time range for this Dashboard in UTC."
```

**Review by lizokm — APPROVED** [review]
Left a few small suggestions, otherwise this looks good!

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 9]
```diff
@@ -44,24 +54,28 @@ def _get_dashboard(self, request: Request, organization, dashboard_id):
...
         """
-        Retrieve an Organization's Dashboard
-        ````````````````````````````````````
-
-        Return details on an individual organization's dashboard.
-
-        :pparam Organization organization: the organization the dashboard belongs to.
-        :pparam Dashboard dashboard: the dashboard object
-        :auth: required
+        Return details on an individual organization's custom dashboard.
```
Comment (lizokm) [inline src/sentry/api/endpoints/organization_dashboard_details.py:78]:
```suggestion
        Return details about an organization's custom dashboard.
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 10]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
 
 class DashboardDetailsSerializer(CamelSnakeSerializer[Dashboard]):
     # Is a string because output serializers also make it a string.
-    id = serializers.CharField(required=False)
-    title = serializers.CharField(required=False, max_length=255)
-    widgets = DashboardWidgetSerializer(many=True, required=False)
-    projects = serializers.ListField(child=serializers.IntegerField(), required=False, default=[])
+    id = serializers.CharField(required=False, help_text="The unique id of the Dashboard.")
+    title = serializers.CharField(
+        required=False, max_length=255, help_text="The user defined title for this Dashboard."
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:452]:
```suggestion
        required=False, max_length=255, help_text="The user-defined dashboard title."
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 11]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
...
+    )
+    widgets = DashboardWidgetSerializer(
+        many=True, required=False, help_text="A json list of widgets saved in this Dashboard."
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:455]:
```suggestion
        many=True, required=False, help_text="A json list of widgets saved in this dashboard."
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 12]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
...
+    )
+    projects = serializers.ListField(
+        child=serializers.IntegerField(),
+        required=False,
+        default=[],
+        help_text="The saved projects filter for this Dashboard.",
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:461]:
```suggestion
        help_text="The saved projects filter for this dashboard.",
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 13]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
...
+    )
     environment = serializers.ListField(
-        child=serializers.CharField(), required=False, allow_null=True
+        child=serializers.CharField(),
+        required=False,
+        allow_null=True,
+        help_text="The saved environment filter for this Dashboard.",
+    )
+    period = serializers.CharField(
+        required=False, allow_null=True, help_text="The saved time range period for this Dashboard."
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:470]:
```suggestion
        required=False, allow_null=True, help_text="The saved time range period for this dashboard."
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 14]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
...
+        default=[],
+        help_text="The saved projects filter for this Dashboard.",
+    )
     environment = serializers.ListField(
-        child=serializers.CharField(), required=False, allow_null=True
+        child=serializers.CharField(),
+        required=False,
+        allow_null=True,
+        help_text="The saved environment filter for this Dashboard.",
+    )
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:468]:
```suggestion
        allow_null=True,
        help_text="The saved environment filter for this dashboard.",
    )
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 15]
```diff
@@ -444,19 +447,40 @@ def validate(self, data):
...
+        required=False, allow_null=True, help_text="The saved time range period for this Dashboard."
+    )
+    start = serializers.DateTimeField(
+        required=False, allow_null=True, help_text="The saved start time for this Dashboard."
+    )
+    end = serializers.DateTimeField(
+        required=False, allow_null=True, help_text="The saved end time for this Dashboard."
+    )
+    filters = serializers.DictField(
+        required=False, help_text="The saved filters for this Dashboard."
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:479]:
```suggestion
    )
    start = serializers.DateTimeField(
        required=False, allow_null=True, help_text="The saved start time for this dashboard."
    )
    end = serializers.DateTimeField(
        required=False, allow_null=True, help_text="The saved end time for this dashboard."
    )
    filters = serializers.DictField(
        required=False, help_text="The saved filters for this dashboard."
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 16]
```diff
@@ -719,7 +743,9 @@ def remove_missing_queries(self, widget_id, keep_ids):
 
 
 class DashboardSerializer(DashboardDetailsSerializer):
-    title = serializers.CharField(required=True, max_length=255)
+    title = serializers.CharField(
+        required=True, max_length=255, help_text="The user defined title for this Dashboard."
+    )
```
Comment (lizokm) [inline src/sentry/api/serializers/rest_framework/dashboard.py:748]:
```suggestion
        required=True, max_length=255, help_text="The user defined title for this dashboard."
    )
```

**Review by lizokm — APPROVED** [review]

**Pull request merged**